File: ForRadiant/MotherGrayGenerator/Form1.vb
Imports System.Text

Public Class Form1
    Private Sub RunButton_Click(sender As Object, e As EventArgs) Handles RunButton.Click
        RunButton.Enabled = False

        If InputTextBox.Text = "" Then
            MessageBox.Show("input files are empty")
            RunButton.Enabled = True
            Exit Sub
        End If
        If OutputTextBox.Text = "" Then
            MessageBox.Show("output folder is empty")
            RunButton.Enabled = True
            Exit Sub
        End If
        If Not IO.Directory.Exists(OutputTextBox.Text) Then
            Try
                IO.Directory.CreateDirectory(OutputTextBox.Text)
            Catch ex As Exception
                MessageBox.Show("error making output folder")
                RunButton.Enabled = True
                Exit Sub
            End Try
        End If

        Dim Files() As String = InputTextBox.Text.Replace(vbCrLf, ",").Split(
                     New String() {","}, StringSplitOptions.RemoveEmptyEntries
                 ).Where(
                     Function(s) Not String.IsNullOrWhiteSpace(s)
                 ).ToArray()

        If Files.Count < 0 Then
            MessageBox.Show("input folder does not contain files")
            RunButton.Enabled = True
            Exit Sub
        End If

        Dim Mother(1169, 2531) As Single
        Dim header(7) As String
        For Each inputfile As String In Files

            Try
                Using fs As New IO.StreamReader(inputfile)
                    'skip 8 header lines, save it if input file is mother green
                    If IO.Path.GetFileNameWithoutExtension(inputfile) = "MOTHER G" Then
                        header(0) = fs.ReadLine()
                        header(1) = fs.ReadLine()
                        header(2) = fs.ReadLine()
                        header(3) = fs.ReadLine()
                        header(4) = fs.ReadLine()
                        header(5) = fs.ReadLine()
                        header(6) = fs.ReadLine()
                        header(7) = fs.ReadLine()
                    Else
                        fs.ReadLine()
                        fs.ReadLine()
                        fs.ReadLine()
                        fs.ReadLine()
                        fs.ReadLine()
                        fs.ReadLine()
                        fs.ReadLine()
                        fs.ReadLine()

                    End If

                    Dim row As Integer = 0
                    Do While Not fs.EndOfStream

                        Dim Data() As String = fs.ReadLine.Split(New String() {ControlChars.Tab}, StringSplitOptions.None)
                        For col As Integer = 0 To Data.Length - 2
                            If IO.Path.GetFileNameWithoutExtension(inputfile) = "MOTHER R" Or IO.Path.GetFileNameWithoutExtension(inputfile) = "MOTHER B" Then
                                Mother(col * 2, row) += Data(col)
                                Mother(col * 2 + 1, row) += Data(col)
                            Else
                                Mother(col, row) += Data(col)
                            End If
                        Next
                        row += 1
                    Loop


                End Using
            Catch ex As Exception
                MessageBox.Show("Error reading input files: " & ex.Message)
            End Try


        Next
        Dim outputfile As String = outputTextBox.Text & "\" & "MOTHER new G.txt"

        Try
            Using sw As IO.StreamWriter = New IO.StreamWriter(IO.File.Open(outputfile, IO.FileMode.Create))

                'write header
                For row As Integer = 0 To header.Length - 1
                    sw.WriteLine(header(row))
                Next

                For row As Integer = 0 To 2531

                    Dim sb As New StringBuilder()

                    For col As Integer = 0 To 1169
                        sb.Append(String.Format("{0}" + vbTab, Mother(col, row)))
                    Next

                    'Remove the last tab from each line
                    sb.Length -= 1
                    sw.WriteLine(sb.ToString())
                Next
            End Using
        Catch ex As Exception
            MessageBox.Show("Error writing output file: " & ex.Message)
            Exit Sub
        End Try
        RunButton.Enabled = True
    End Sub

    Private Sub InputTextBox_DragEnter(sender As Object, e As DragEventArgs) Handles inputTextBox.DragEnter
        e.Effect = DragDropEffects.All
        Dim files() As String = e.Data.GetData(DataFormats.FileDrop)
        For Each path In files
            inputTextBox.Text += path + vbCrLf
        Next
        outputTextBox.Text = IO.Path.GetDirectoryName(files(1))

    End Sub

End Class

File: ForRadiant/FileHide/Form1.cs
using System;
using System.IO;
using System.Linq;
using System.Windows.Forms;

namespace FileHide
{
    public partial class Form1 : Form
    {
        private const string controlValuesFileName = "settings.txt";
        private const string logFileName = "filehidelog.txt";
        private const int maxLogLines = 10000; // Maximum number of lines in the log file

        public Form1()
        {
            InitializeComponent();
            LoadControlValues(); // Load control values when the form loads
        }

        private void LoadControlValues()
        {
            if (File.Exists(controlValuesFileName))
            {
                string[] lines = File.ReadAllLines(controlValuesFileName);
                FilesTextBox.Clear();
                FoldersTextBox.Clear();
                ExtensionsTextBox.Clear();

                bool readingFiles = false;
                bool readingFolders = false;
                bool readingExtensions = false;

                foreach (string line in lines)
                {
                    if (line == "[Files]")
                    {
                        readingFiles = true;
                        readingFolders = false;
                        readingExtensions = false;
                        continue;
                    }
                    if (line == "[Folders]")
                    {
                        readingFiles = false;
                        readingFolders = true;
                        readingExtensions = false;
                        continue;
                    }
                    if (line == "[Extensions]")
                    {
                        readingFiles = false;
                        readingFolders = false;
                        readingExtensions = true;
                        continue;
                    }

                    if (readingFiles)
                    {
                        FilesTextBox.AppendText(line + Environment.NewLine);
                    }
                    if (readingFolders)
                    {
                        FoldersTextBox.AppendText(line + Environment.NewLine);
                    }
                    if (readingExtensions)
                    {
                        ExtensionsTextBox.AppendText(line);
                    }
                }
            }
        }

        private void SaveControlValues()
        {
            using (StreamWriter sw = new StreamWriter(controlValuesFileName))
            {
                sw.WriteLine("[Files]");
                foreach (string line in FilesTextBox.Lines)
                {
                    sw.WriteLine(line);
                }

                sw.WriteLine("[Folders]");
                foreach (string line in FoldersTextBox.Lines)
                {
                    sw.WriteLine(line);
                }

                sw.WriteLine("[Extensions]");
                sw.WriteLine(ExtensionsTextBox.Text);
            }
        }

        private void Form1_FormClosing(object sender, FormClosingEventArgs e)
        {
            SaveControlValues(); // Save control values when the form is closing
        }

        private void HideButton_Click(object sender, EventArgs e)
        {
            Password passwordForm = new Password();
            passwordForm.ShowDialog();

            if (!passwordForm.IsPasswordCorrect)
            {
                WriteToLog("Incorrect password. Operation canceled.");
                return;
            }
            RemoveEmptyLines(FilesTextBox);
            RemoveEmptyLines(FoldersTextBox);
            HideUnhideFiles(true);
        }

        private void UnhideButton_Click(object sender, EventArgs e)
        {
            Password passwordForm = new Password();
            passwordForm.ShowDialog();

            if (!passwordForm.IsPasswordCorrect)
            {
                WriteToLog("Incorrect password. Operation canceled.");
                return;
            }
            RemoveEmptyLines(FilesTextBox);
            RemoveEmptyLines(FoldersTextBox);
            HideUnhideFiles(false);
        }

        private void HideUnhideFiles(bool hideFiles)
        {
            string[] filePaths = FilesTextBox.Lines;
            string[] folderPaths = FoldersTextBox.Lines;
            string[] extensions = ExtensionsTextBox.Text.Split(new[] { ',' }, StringSplitOptions.RemoveEmptyEntries)
                                                          .Select(ext => ext.Trim().ToLower()).ToArray();

            bool includeAllFiles = extensions.Contains("*");

            // Process individual files
            foreach (string filePath in filePaths)
            {
                ProcessFile(filePath, hideFiles);
            }

            // Process folders
            foreach (string folderPath in folderPaths)
            {
                if (string.IsNullOrWhiteSpace(folderPath) || !Directory.Exists(folderPath))
                {
                    WriteToLog($"Folder '{folderPath}' does not exist.");
                    continue;
                }

                ProcessFolder(folderPath, extensions, includeAllFiles, hideFiles);
            }
        }

        private void ProcessFile(string filePath, bool hideFiles)
        {
            try
            {
                if (string.IsNullOrWhiteSpace(filePath) || !File.Exists(filePath))
                {
                    WriteToLog($"File '{filePath}' does not exist.");
                    return;
                }

                FileAttributes attributes = File.GetAttributes(filePath);

                if (hideFiles)
                {
                    // Hide the file
                    attributes |= FileAttributes.Hidden | FileAttributes.System;
                    File.SetAttributes(filePath, attributes);
                    WriteToLog($"File '{filePath}' is now hidden.");
                }
                else
                {
                    // Unhide the file
                    attributes &= ~FileAttributes.Hidden;
                    attributes &= ~FileAttributes.System;
                    File.SetAttributes(filePath, attributes);
                    WriteToLog($"File '{filePath}' is now visible.");
                }
            }
            catch (Exception ex)
            {
                WriteToLog($"Error processing file '{filePath}': {ex.Message}");
            }
        }

        private void ProcessFolder(string folderPath, string[] extensions, bool includeAllFiles, bool hideFiles)
        {
            try
            {
                if (includeAllFiles)
                {
                    extensions = new string[] { "*.*" };
                }

                foreach (string extension in extensions)
                {
                    string searchPattern = includeAllFiles ? "*.*" : $"*.{extension}";
                    string[] files = Directory.GetFiles(folderPath, searchPattern, SearchOption.AllDirectories);

                    foreach (string file in files)
                    {
                        ProcessFile(file, hideFiles);
                    }
                }
            }
            catch (Exception ex)
            {
                WriteToLog($"Error processing folder '{folderPath}': {ex.Message}");
            }
        }

        private void FilesTextBox_DragEnter(object sender, DragEventArgs e)
        {
            if (e.Data.GetDataPresent(DataFormats.FileDrop))
            {
                e.Effect = DragDropEffects.Copy;
            }
            else
            {
                e.Effect = DragDropEffects.None;
            }
        }

        private void FilesTextBox_DragDrop(object sender, DragEventArgs e)
        {
            if (e.Data.GetDataPresent(DataFormats.FileDrop))
            {
                string[] files = (string[])e.Data.GetData(DataFormats.FileDrop);
                foreach (string file in files)
                {
                    if (!FilesTextBox.Lines.Contains(file))
                    {
                        FilesTextBox.AppendText(file + Environment.NewLine);
                    }
                }
            }
        }

        private void FoldersTextBox_DragEnter(object sender, DragEventArgs e)
        {
            if (e.Data.GetDataPresent(DataFormats.FileDrop))
            {
                e.Effect = DragDropEffects.Copy;
            }
            else
            {
                e.Effect = DragDropEffects.None;
            }
        }

        private void FoldersTextBox_DragDrop(object sender, DragEventArgs e)
        {
            if (e.Data.GetDataPresent(DataFormats.FileDrop))
            {
                string[] folders = (string[])e.Data.GetData(DataFormats.FileDrop);
                foreach (string folder in folders)
                {
                    if (!FoldersTextBox.Lines.Contains(folder))
                    {
                        FoldersTextBox.AppendText(folder + Environment.NewLine);
                    }
                }
            }
        }

        private void RemoveEmptyLines(TextBox textBox)
        {
            var lines = textBox.Lines
                .Where((line, index) => !string.IsNullOrWhiteSpace(line) || index == textBox.Lines.Length - 1)
                .ToArray();
            textBox.Lines = lines;
        }

        private void WriteToLog(string message)
        {
            try
            {
                // Trim log file if it exceeds maximum lines
                if (File.Exists(logFileName))
                {
                    var logLines = File.ReadAllLines(logFileName).ToList();
                    if (logLines.Count >= maxLogLines)
                    {
                        logLines = logLines.Skip(logLines.Count - maxLogLines + 1).ToList();
                    }
                    logLines.Add($"{DateTime.Now}: {message}");
                    File.WriteAllLines(logFileName, logLines);
                }
                else
                {
                    File.WriteAllText(logFileName, $"{DateTime.Now}: {message}\n");
                }
            }
            catch (Exception ex)
            {
                // If logging fails, you might want to handle this accordingly
                MessageBox.Show($"Error writing to log file: {ex.Message}");
            }
        }

        private void RemoveEmptyLinesButton_Click(object sender, EventArgs e)
        {
            RemoveEmptyLines(FilesTextBox);
            RemoveEmptyLines(FoldersTextBox);
        }
    }
}
